rfcs/0079-enforce-prs.md

```diff
@@ -0,0 +1,80 @@
+---
+feature: No more direct pushes to master and release branches
+start-date: 2020-09-25
+author: Janne Heß <janne@hess.ooo>
+co-authors: Matthias Beyer <mail@beyermatthias.de>
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Enforce usage of Pull Requests for **all** contributions to nixpkgs master and release branches, which implies that these branches do not allow direct pushes anymore.
+
+This implements the four-eyes principle and allows easier change discussion, both before and after the merge, and improves overall security since each Pull Request has to be approved by at least one other person prior to merging.
```

Comment:
> Pull Request has to be approved by at least one other person prior to merging.

This was discussed on the merge bot RFC as well. The group of people with push rights is too small. There we proposed introducing a new group, trusted reviewers, that could approve and then merge via the bot.

Reply:
I don't think requiring 4-eye-review for stable branches will in fact improve the situation there at all. From subjective observation there appear to be only at most a hand full of people caring for a released version of NixOS besides some random back ports and high visibility issues. During the 19.09 & 20.03 release cycles there were multiple cases were PRs to fix security related stuff had been open for days. If I wouldn't be able to merge my own PRs (even without CI as they might be rebuilding everything due to touching OpenSSL) I'd probably not have done that work (and they might still be open until today).

We can probably discuss how useful protecting the branch against (fast-forward) pushes is but the review requirement should probably get a dedicated RFC (and discussion).

Reply:
> The group of people with push rights is too small.

No, it is too big. That's the whole point of this RFC - too many people can break master/release and push security issues.

Reply:
We do not have enough people with merge access to merge all of the slightly lower profile security fixes in a timely fashion. (Also, we lack enough people to triage security and general severity of bugfix releases, so we need to somehow speed up merging almost all of the minor updates).

Tricking people into merging non-obvious security issues will only become easier by increasing the amount of rote process.

Reply:
Agreed. I don't see any conflicting points between your statement and mine. :smile:

Reply:
If we do not have enough throughput, putting speedbumps to workflows of people providing the throughput is a bad idea. This proposal is doing that.